libraries/AP_Camera/AP_Camera_MAVLinkCamV2.cpp

```diff
@@ -0,0 +1,215 @@
+    mavlink_command_long_t pkt {};
+    pkt.command = MAV_CMD_SET_CAMERA_ZOOM;
+    pkt.param1 = ZOOM_TYPE_CONTINUOUS;  // Zoom Type, 0:ZOOM_TYPE_STEP, 1:ZOOM_TYPE_CONTINUOUS, 2:ZOOM_TYPE_RANGE, 3:ZOOM_TYPE_FOCAL_LENGTH
+    pkt.param2 = zoom_step;             // Zoom Value
+
+    _link->send_message(MAVLINK_MSG_ID_COMMAND_LONG, (const char*)&pkt);
+
+    return true;
+}
+
+// set focus in, out or hold.  returns true on success
+// focus in = -1, focus hold = 0, focus out = 1
+bool AP_Camera_MAVLinkCamV2::set_manual_focus_step(int8_t focus_step)
+{
+    if (_link == nullptr) {
+        return false;
+    }
+
+    // prepare and send message
+    mavlink_command_long_t pkt {};
+    pkt.command = MAV_CMD_SET_CAMERA_FOCUS;
+    pkt.param1 = FOCUS_TYPE_CONTINUOUS; // Focus Type, 0:FOCUS_TYPE_STEP, 1:FOCUS_TYPE_CONTINUOUS, 2:FOCUS_TYPE_RANGE, 3:FOCUS_TYPE_METERS, 4:FOCUS_TYPE_AUTO, 5:FOCUS_TYPE_AUTO_SINGLE, 5:FOCUS_TYPE_AUTO_CONTINUOUS
+    pkt.param2 = focus_step;            // Focus Value
+
+    _link->send_message(MAVLINK_MSG_ID_COMMAND_LONG, (const char*)&pkt);
+
+    return true;
+}
+
+// auto focus.  returns true on success
+bool AP_Camera_MAVLinkCamV2::set_auto_focus()
+{
+    if (_link == nullptr) {
+        return false;
+    }
+
+    // prepare and send message
+    mavlink_command_long_t pkt {};
+    pkt.command = MAV_CMD_SET_CAMERA_FOCUS;
+    pkt.param1 = FOCUS_TYPE_AUTO;   // Focus Type, 0:FOCUS_TYPE_STEP, 1:FOCUS_TYPE_CONTINUOUS, 2:FOCUS_TYPE_RANGE, 3:FOCUS_TYPE_METERS, 4:FOCUS_TYPE_AUTO, 5:FOCUS_TYPE_AUTO_SINGLE, 5:FOCUS_TYPE_AUTO_CONTINUOUS
+    pkt.param2 = 0;                 // Focus Value
+
+    _link->send_message(MAVLINK_MSG_ID_COMMAND_LONG, (const char*)&pkt);
+
+    return true;
+}
+
+// handle incoming mavlink message including CAMERA_INFORMATION
+void AP_Camera_MAVLinkCamV2::handle_message(mavlink_channel_t chan, const mavlink_message_t &msg)
+{
+    // exit immediately if this is not our message
+    if (msg.sysid != _sysid || msg.compid != _compid) {
+        return;
+    }
+
+    // handle CAMERA_INFORMATION
+    if (msg.msgid == MAVLINK_MSG_ID_CAMERA_INFORMATION) {
+        mavlink_camera_information_t cam_info;
+        mavlink_msg_camera_information_decode(&msg, &cam_info);
+
+        const uint8_t fw_ver_major = cam_info.firmware_version & 0x000000FF;
+        const uint8_t fw_ver_minor = (cam_info.firmware_version & 0x0000FF00) >> 8;
+        const uint8_t fw_ver_revision = (cam_info.firmware_version & 0x00FF0000) >> 16;
+        const uint8_t fw_ver_build = (cam_info.firmware_version & 0xFF000000) >> 24;
+
+        // display camera info to user
+        gcs().send_text(MAV_SEVERITY_INFO, "Camera: %s %s fw:%u.%u.%u.%u",
+                cam_info.vendor_name,
+                cam_info.model_name,
+                (unsigned)fw_ver_major,
+                (unsigned)fw_ver_minor,
+                (unsigned)fw_ver_revision,
+                (unsigned)fw_ver_build);
+
+        _got_camera_info = true;
```

Comment:
_camera_cap_flags = cam_info.flags; (_camera_cap_flags needs to be defined in the class)

Reply:
Txs for the review, I've added basic capability flags support.